# [9장 1강] 내장 함수 — 최종 실습문제

## 실습 목표
- `isinstance()`, `all()`, `any()`를 활용해 데이터의 자료형과 조건 충족 여부를 검증할 수 있다.
- `sum()`, `len()`, `round()`, `min()`, `max()`로 수치 데이터를 요약할 수 있다.
- `map()`, `filter()`를 활용해 여러 데이터를 일괄 변환·추출할 수 있다.
- `sorted()`, `zip()`, `enumerate()`로 데이터를 정렬·결합하고 순번을 부여할 수 있다.
- 빈 데이터에서 `all()`과 `any()`의 동작 차이를 이해할 수 있다.

## 실습 환경
- Python 3.14
- VS Code 또는 Jupyter Notebook
- 외부 라이브러리 없음
- 표준 라이브러리 `statistics`는 심화 문제에서만 사용

## 실무
회원가입 검증, 로그인 상태 확인, 점수·매출 요약, 문자열 데이터 변환, 대상자 목록 매칭처럼 실무에서 반복되는 작업은 파이썬 내장 함수를 조합하면 간결하게 처리할 수 있습니다.

## 진행 방식
- 각 문제에서 지정한 내장 함수를 반드시 사용하세요.
- 같은 결과가 나오더라도 이번 차시에서는 해당 함수의 역할을 이해하는 데 집중하세요.
- `map()`과 `filter()`의 결과는 필요할 때 `list()`로 변환해 확인하세요.
- `statistics.mean()`과 `statistics.median()`은 파이썬 **내장 함수가 아니라 표준 라이브러리 함수**입니다.

---

## 필수 1. 회원가입 및 로그인 상태 유효성 검증

### 문제 1-1. 자료형·전체 조건·하나 이상의 조건 확인

### 요구사항
1. `age = 25`를 작성하세요.
2. `isinstance(age, int)`로 정수 자료형인지 확인하세요.
3. 다음 필수 약관 리스트를 작성하세요.

```python
required_agreements = [True, True, True]
```

4. `all()`을 사용해 모든 필수 약관에 동의했는지 확인하세요.
5. 다음 로그인 연동 상태를 작성하세요.

```python
login_status = [False, True, False]
```

6. `any()`를 사용해 로그인 수단이 하나라도 활성화되어 있는지 확인하세요.
7. 세 결과를 각각 출력하세요.

In [6]:
# isinstance(), all(), any()를 사용해 데이터를 검증하세요.
age = '25'
print('나이 자료형 확인:', isinstance(age, int))

required_agreements = [True, False, True]
print('필수 약관 전체 동의', all(required_agreements))

login_status = [False, False, False]
print('로그인 수단 활성화 여부:', any(login_status))

나이 자료형 확인: False
필수 약관 전체 동의 False
로그인 수단 활성화 여부: False


### 예상 결과

```text
나이 자료형 확인: True
필수 약관 전체 동의: True
로그인 수단 활성화 여부: True
```

### 추가 확인
다음 값으로 바꿔 결과 차이를 확인하세요.

```python
age = "25"
required_agreements = [True, False, True]
login_status = [False, False, False]
```

---

## 필수 2. 점수 데이터 요약

### 문제 2-1. 평균·최저점·최고점 계산

### 데이터

```python
scores = [85, 62, 92, 78, 55]
```

### 요구사항
1. `sum()`과 `len()`을 사용해 평균을 계산하세요.
2. `round(..., 2)`로 평균을 소수점 둘째 자리까지 반올림하세요.
3. `min()`으로 최저점을 구하세요.
4. `max()`로 최고점을 구하세요.
5. 전체 점수 개수도 `len()`으로 출력하세요.

In [9]:
# 점수 개수, 평균, 최저점, 최고점을 계산하세요.
scores = [85, 62, 92, 78, 55]

average_score = sum(scores) / len(scores)
print('점수 개수:', len(scores))
print('평균 점수:', round(average_score, 2))
print('최저점:', min(scores))
print('최고점:', max(scores))

점수 개수: 5
평균 점수: 74.4
최저점: 55
최고점: 92


### 예상 결과

```text
점수 개수: 5
평균 점수: 74.4
최저점: 55
최고점: 92
```

> `round(value, 2)`는 둘째 자리까지 반올림하지만 항상 `74.40`처럼 두 자리 형식으로 표시되는 것은 아닙니다.

---

## 필수 3. `map()`과 `filter()`를 활용한 데이터 가공

### 문제 3-1. 가격 변환과 합격 점수 추출

### 데이터 A

```python
price_texts = ["12000", "25000", "31000"]
```

### 데이터 B

```python
scores = [45, 80, 55, 90, 30]
```

### 요구사항
1. `map()`과 `int`를 사용해 가격 문자열을 모두 정수로 변환하세요.
2. 결과를 `list()`로 변환하세요.
3. `sum()`으로 총 가격을 계산하세요.
4. `filter()`와 `lambda`를 사용해 60점 이상 점수만 추출하세요.
5. 결과를 `list()`로 변환하세요.

In [16]:
# map()으로 가격을 변환하고 filter()로 합격 점수를 추출하세요.
price_texts = ['12000', '25000', '31000']

price = list(map(int, price_texts))
print('반환된 가격:', price)

total_price = sum(price)
print('총 가격:', total_price)

scores = [45, 80, 55, 90, 30]
high_score = list(filter(lambda x: x >= 60, scores))
print('합격 점수:', high_score)


반환된 가격: [12000, 25000, 31000]
총 가격: 68000
합격 점수: [80, 90]


### 예상 결과

```text
변환된 가격: [12000, 25000, 31000]
총 가격: 68000
합격 점수: [80, 90]
```

---

## 필수 4. 정렬과 최고·최저 데이터 확인

### 문제 4-1. 점수 정렬

### 데이터

```python
scores = [85, 95, 90, 78]
```

### 요구사항
1. `sorted(scores)`로 오름차순 정렬 결과를 만드세요.
2. `sorted(scores, reverse=True)`로 내림차순 정렬 결과를 만드세요.
3. `min()`과 `max()`로 최저점·최고점을 확인하세요.
4. 원본 `scores`도 다시 출력하여 `sorted()`가 새로운 리스트를 반환하는지 확인하세요.

In [20]:
# sorted(), min(), max()를 사용해 점수 데이터를 확인하세요.
scores = [85, 95, 90, 78]

print('오름차순:', sorted(scores))
print('내림차순:', sorted(scores, reverse=True))
print('최저점:', min(scores))
print('최고점:', max(scores))
print('원본:', scores)

오름차순: [78, 85, 90, 95]
내림차순: [95, 90, 85, 78]
최저점: 78
최고점: 95
원본: [85, 95, 90, 78]


### 예상 결과

```text
오름차순: [78, 85, 90, 95]
내림차순: [95, 90, 85, 78]
최저점: 78
최고점: 95
원본: [85, 95, 90, 78]
```

---

## 심화 1. `zip()`과 `enumerate()`를 활용한 대상자 명부 생성

### 문제 5-1. 이벤트 당첨자 순번과 이메일 매칭

### 데이터

```python
names = ["이민서", "박준형"]
emails = ["minseo@ssu.ac.kr", "junhyung@daum.net"]
```

### 요구사항
1. `zip()`으로 이름과 이메일을 같은 순서끼리 묶으세요.
2. `enumerate(..., start=1)`로 번호를 1부터 부여하세요.
3. `for` 반복문으로 다음 형식으로 출력하세요.

```text
1번 당첨자: 이민서 | 이메일: minseo@ssu.ac.kr
2번 당첨자: 박준형 | 이메일: junhyung@daum.net
```

In [22]:
# zip()과 enumerate()를 사용해 당첨자 명부를 출력하세요.
names = ["이민서", "박준형"]
emails = ["minseo@ssu.ac.kr", "junhyung@daum.net"]

for number, (name, email) in enumerate(zip(names, emails), start=1):
    print(f'{number}번 당첨자: {name} | 이메일: {email}')


1번 당첨자: 이민서 | 이메일: minseo@ssu.ac.kr
2번 당첨자: 박준형 | 이메일: junhyung@daum.net


In [24]:
names = ['철수', '영희', '민수']
scores = [80, 95, 70]

for name, score in zip(names, scores):
    print(f'{name}: {score}')

철수: 80
영희: 95
민수: 70


In [28]:
fruits = ['사과', '바나나', '딸기']
for number, fruits in enumerate(fruits, start =1):
    print(f'{number}번 과일: {fruits}')

1번 과일: 사과
2번 과일: 바나나
3번 과일: 딸기


---

## 심화 2. 빈 리스트에서 `all()`과 `any()` 주의

### 문제 6-1. 필수 동의 데이터 누락 검증

### 요구사항
1. 빈 리스트를 작성하세요.

```python
user_inputs = []
```

2. `all(user_inputs)` 결과를 출력하세요.
3. `any(user_inputs)` 결과를 출력하세요.
4. 다음 두 조건문의 차이를 확인하세요.

```python
if all(user_inputs):
    print("검증 통과")

if user_inputs and all(user_inputs):
    print("안전하게 검증 통과")
else:
    print("입력 데이터가 없거나 조건에 맞지 않습니다.")
```

5. 빈 리스트에서 `all()`이 `True`가 되는 이유와 실무에서의 주의점을 작성하세요.

In [ ]:
# 빈 리스트에서 all()과 any()의 결과를 확인하세요.
user_inputs = []

print(all(user_inputs)) #요소 중 단 하나라도 False라면 False
print(any(user_inputs)) #요소 중 단 하나라도 True라면 True

if all(user_inputs):
    print('검증 통과')

if user_inputs and all(user_inputs):
    print('안전하게 검증 통과')
else:
    print('입력 데이터가 없거나 조건에 맞지 않습니다.')


True
False
검증 통과
입력 데이터가 없거나 조건에 맞지 않습니다.


### 확인 사항

```text
all([]) → True
any([]) → False
```

필수 데이터 목록 자체가 누락될 수 있는 상황이라면 `all()`만 확인하지 말고 데이터가 실제로 존재하는지도 함께 확인해야 합니다.

---

## 심화 3. 평균과 중앙값 비교

### 문제 7-1. 이상치가 포함된 자산 데이터 요약

> 이 문제에서 사용하는 `statistics.mean()`과 `statistics.median()`은 내장 함수가 아니라 파이썬 표준 라이브러리 함수입니다.

### 데이터

```python
assets = [10, 20, 15, 30, 1000000000]
```

### 요구사항
1. `statistics` 모듈을 불러오세요.
2. `statistics.mean()`으로 평균을 구하세요.
3. `statistics.median()`으로 중앙값을 구하세요.
4. 두 결과를 비교하세요.
5. 극단적으로 큰 값 하나가 평균과 중앙값에 각각 어떤 영향을 주는지 작성하세요.

In [34]:
# statistics.mean()과 statistics.median()을 비교하세요.
import statistics

assets = [10, 2, 15, 30, 1000000000]

print('평균:', statistics.mean(assets))
print('중앙값:', statistics.median(assets))


평균: 200000011.4
중앙값: 15


---

## 실습 마무리 질문

1. `isinstance()`는 어떤 상황에서 유용한가요?
2. `all()`과 `any()`의 차이는 무엇인가요?
3. 왜 `all([])`은 `True`인가요?
4. 평균을 구할 때 `sum()`과 `len()`은 각각 어떤 역할을 하나요?
5. `map()`과 `filter()`의 역할 차이는 무엇인가요?
6. `sorted()`와 원본 리스트의 관계는 무엇인가요?
7. `enumerate(start=1)`은 어떤 작업을 편리하게 하나요?
8. `zip()`에서 두 리스트의 길이가 다르면 어떻게 되나요?
9. 평균과 중앙값 중 이상치의 영향을 더 적게 받는 값은 무엇인가요?

### 답변 작성

1. 
2. 
3. 
4. 
5. 
6. 
7. 
8. 
9.

---

## 제출 전 체크리스트

- [ ] `isinstance()`로 자료형을 확인했다.
- [ ] `all()`과 `any()`의 차이를 확인했다.
- [ ] `sum()`, `len()`, `round()`, `min()`, `max()`를 사용했다.
- [ ] `map()`으로 여러 값을 일괄 변환했다.
- [ ] `filter()`로 조건에 맞는 값을 추출했다.
- [ ] `sorted()`로 오름차순·내림차순 정렬을 수행했다.
- [ ] `zip()`으로 여러 목록을 결합했다.
- [ ] `enumerate(start=1)`로 순번을 부여했다.
- [ ] 빈 리스트에서 `all()`과 `any()`의 동작을 확인했다.
- [ ] `statistics.mean()`과 `median()`의 차이를 확인했다.